worlds/dkc3/Client.py

```diff
@@ -0,0 +1,205 @@
+from SNIClient import Context, snes_buffered_write, snes_flush_writes, snes_read
+from Patch import GAME_DKC3
+
+snes_logger = logging.getLogger("SNES")
+
+# DKC3 - DKC3_TODO: Check these values
+ROM_START = 0x000000
+WRAM_START = 0xF50000
+WRAM_SIZE = 0x20000
+SRAM_START = 0xE00000
+
+SAVEDATA_START = WRAM_START + 0xF000
+SAVEDATA_SIZE = 0x500
+
+DKC3_ROMNAME_START = 0x00FFC0
+DKC3_ROMHASH_START = 0x7FC0
+ROMNAME_SIZE = 0x15
+ROMHASH_SIZE = 0x15
+
+DKC3_RECV_PROGRESS_ADDR = WRAM_START + 0x632     # DKC3_TODO: Find a permanent home for this
+DKC3_FILE_NAME_ADDR = WRAM_START + 0x5D9
+DEATH_LINK_ACTIVE_ADDR = DKC3_ROMNAME_START + 0x15     # DKC3_TODO: Find a permanent home for this
+
+
+async def deathlink_kill_player(ctx: Context):
+    pass
+    #if ctx.game == GAME_DKC3:
+        # DKC3_TODO: Handle Receiving Deathlink
+
+
+async def dkc3_rom_init(ctx: Context):
+    if not ctx.rom:
+        ctx.finished_game = False
+        ctx.death_link_allow_survive = False
+        game_name = await snes_read(ctx, DKC3_ROMNAME_START, 0x15)
+        if game_name is None or game_name != b"DONKEY KONG COUNTRY 3":
+            return False
+        else:
+            ctx.game = GAME_DKC3
+            ctx.items_handling = 0b111  # remote items
+
+        rom = await snes_read(ctx, DKC3_ROMHASH_START, ROMHASH_SIZE)
+        if rom is None or rom == bytes([0] * ROMHASH_SIZE):
+            return False
+
+        ctx.rom = rom
+
+        #death_link = await snes_read(ctx, DEATH_LINK_ACTIVE_ADDR, 1)
+        ## DKC3_TODO: Handle Deathlink
+        #if death_link:
+        #    ctx.allow_collect = bool(death_link[0] & 0b100)
+        #    await ctx.update_death_link(bool(death_link[0] & 0b1))
+    return True
+
+
+async def dkc3_game_watcher(ctx: Context):
+    if ctx.game == GAME_DKC3:
+        # DKC3_TODO: Handle Deathlink
+        save_file_name = await snes_read(ctx, DKC3_FILE_NAME_ADDR, 0x5)
+        if save_file_name is None or save_file_name[0] == 0x00:
+            # We haven't loaded a save file
+            return
+
+        from worlds.dkc3.Rom import location_rom_data, item_rom_data
+        for loc_id, loc_data in location_rom_data.items():
+            if loc_id not in ctx.locations_checked:
+                data = await snes_read(ctx, WRAM_START + loc_data[0], 1)
+                masked_data = data[0] & (1 << loc_data[1])
+                bit_set = (masked_data != 0)
+                invert_bit = ((len(loc_data) >= 3) and loc_data[2])
+                if bit_set != invert_bit:
+                    # DKC3_TODO: Handle non-included checks
+                    ctx.locations_checked.add(loc_id)
+                    location = ctx.location_names[loc_id]
+                    snes_logger.info(
+                        f'New Check: {location} ({len(ctx.locations_checked)}/{len(ctx.missing_locations) + len(ctx.checked_locations)})')
+                    await ctx.send_msgs([{"cmd": 'LocationChecks', "locations": [loc_id]}])
```

Comment:
to avoid bogus sends, ideally you
1. collect all changed locations
2. revalidate that the game is still active, ideally
   * re-validate ROM (or seed or slot or whatever)
   * check "is in game" flag (i.e. not sitting in the menu)
3. then send all collected locations

Even more robust would be to repeat the check from 2. as 0., and compare an int between those two, so that you can verify data is consistent for all of 1.

Reply:
I've added some amount of safeguarding along the lines of what you were describing. I would appreciate it if you could look over what I did as a sanity check (I did functionally test it, and it seems to work).

Reply:
Looks good to me for basic protection. It's probably good enough for day1, but please still read below. If DKC3 becomes super popular, failed sends may still creep up.

The last suggestion i had there could be implemented by
* read the save_file_name (this would be added)
* read all the locations, build the list (as it is now)
* read the save_file_name again (as it is now)
* compare the first save_file_name to the second (this would be added)

However, doing this blindly does not guarantee fixing everything because when you reset an fxpak back to the menu, the ROM, WRAM and SRAM is still available, but the m3nu.bin will overwrite *some* WRAM locations. This means you'd have to validate that either no locations are being overwritten by m3nu, or the save_file_name in WRAM is overwritten as well.